### PR TITLE
Adding the FPS-limit and Vsync options

### PR DESCRIPTION
**Pull request by AnonFS**
Addresses the https://github.com/Vita3K/Vita3K/issues/367 issue.

### CODE REVIEW

**Comment by 1whatleytay** [comment]
Awesome. What's with these guys?
Screen Shot 2019-06-25 at 5 01 47 PM

**Comment by AnonFS** [comment]
Error slipped in. How would I go about getting these in order?

**Comment by Zangetsu38** [comment]
Desired fps is for what ? and for me only need vsync option and fps limit

**Comment by Zangetsu38** [comment]
ah for glbinding you can maybe remove change, one pr is opened for delete it
https://github.com/Vita3K/Vita3K/pull/527

**Comment by AnonFS** [comment]
Desired FPS is just the name for the constant you want to use as your limit. FPS-Limit itself is just a bool deciding if the limit gets applied or not. I tried to stick with the established naming scheme here, trivial to change if you guys think it could use a better name.

File: src/emulator/main.cpp [hunk 1]
```diff
@@ -154,9 +159,19 @@ int main(int argc, char *argv[]) {
...
+        if (!host.cfg.sync_rendering) {
             host.display.condvar.notify_all();
-        else {
+            if (host.cfg.fps_limit) {
+                const std::chrono::microseconds TARGET_TIME{ static_cast<uint32_t>(1000000.0 / (host.cfg.desired_fps+1)) };
+                
+                frame_end = std::chrono::steady_clock::now();
+                std::chrono::microseconds frame_processing_time = std::chrono::duration_cast<std::chrono::microseconds>(frame_end - frame_start);
+                if (frame_processing_time < TARGET_TIME) {
+			        std::chrono::microseconds frame_idle_time = TARGET_TIME - frame_processing_time;
```
Comment (pent0) [inline src/emulator/main.cpp]: Mixed tabs and spaces

File: src/emulator/main.cpp [hunk 2]
```diff
@@ -154,9 +159,19 @@ int main(int argc, char *argv[]) {
         }
         gui::draw_end(host.window.get());
 
-        if (!host.cfg.sync_rendering)
+        if (!host.cfg.sync_rendering) {
             host.display.condvar.notify_all();
-        else {
+            if (host.cfg.fps_limit) {
+                const std::chrono::microseconds TARGET_TIME{ static_cast<uint32_t>(1000000.0 / (host.cfg.desired_fps+1)) };
```
Comment (pent0) [inline src/emulator/main.cpp]:
```suggestion
                const std::chrono::microseconds TARGET_TIME{ static_cast<uint32_t>(1000000.0 / (host.cfg.desired_fps + 1)) };
```

This is not really necessary but it looks more nice. Also it's more great if you can name the 1000000.0 (to an constexpr maybe? for future in case anyone want to understand the code)

**Comment by pent0** [comment]
Please fix the conflicts.

File: src/emulator/config/include/config/config.h [hunk 3]
```diff
@@ -43,10 +43,13 @@ struct Config {
...
     bool texture_cache = true;
-    int sys_button = 1;
-    int sys_lang = 1;
+    bool fps_limit = false;
+    bool wait_for_vsync = true;
     std::string background_image = {};
     float background_alpha = 0.300f;
+    int desired_fps = 60;
+    int sys_button = 1;
+    int sys_lang = 1;
```
Comment (totlmstr) [inline src/emulator/config/include/config/config.h]: Revert the changes in `config.h` and instead move your new values underneath `discord_rich_presence` like you have it in `config.cpp`. `config.h` mimics the placement of `config.cpp` for consistency.

File: src/emulator/config/src/config.cpp [hunk 4]
```diff
@@ -129,6 +129,9 @@ static ExitCode parse(Config &cfg, const fs::path &load_path, const std::string
     get_yaml_value(config_node, "log-level", &cfg.log_level, static_cast<int>(spdlog::level::trace));
     get_yaml_value(config_node, "pref-path", &cfg.pref_path, root_pref_path);
     get_yaml_value(config_node, "discord-rich-presence", &cfg.discord_rich_presence, true);
+    get_yaml_value(config_node, "fps-limit", &cfg.fps_limit, false);
+    get_yaml_value(config_node, "desired-fps", &cfg.desired_fps, static_cast<int>(60));
+    get_yaml_value(config_node, "wait-for-vsync", &cfg.wait_for_vsync, true);
```
Comment (totlmstr) [inline src/emulator/config/src/config.cpp]: Don't forget to add in these values for `serialize_config` and `merge_configs`.

File: src/emulator/config/src/config.cpp [hunk 5]
```diff
@@ -180,6 +183,9 @@ ExitCode serialize_config(Config &cfg, const fs::path &output_path) {
     config_file_emit_single(emitter, "pref-path", cfg.pref_path);
     config_file_emit_vector(emitter, "lle-modules", cfg.lle_modules);
     config_file_emit_single(emitter, "discord-rich-presence", cfg.discord_rich_presence);
+    config_file_emit_single(emitter, "fps-limit", cfg.fps_limit;
```
Comment (1whatleytay) [inline src/emulator/config/src/config.cpp]:
```suggestion
    config_file_emit_single(emitter, "fps-limit", cfg.fps_limit);
```
The `)` ran away.

Comment (AnonFS) [reply src/emulator/config/src/config.cpp]: Thanks, forgot to check if the build passed on this.
As far as the sleep goes, it works perfectly fine in the homebrews I've tried. Still, might hit up the guys trying out various games to try and use the option for a bit.

File: src/emulator/config/src/config.cpp [hunk 6]
```diff
@@ -180,6 +183,9 @@ ExitCode serialize_config(Config &cfg, const fs::path &output_path) {
     config_file_emit_single(emitter, "pref-path", cfg.pref_path);
     config_file_emit_vector(emitter, "lle-modules", cfg.lle_modules);
     config_file_emit_single(emitter, "discord-rich-presence", cfg.discord_rich_presence);
+    config_file_emit_single(emitter, "fps-limit", cfg.fps_limit);
+    config_file_emit_single(emitter, "desired-fps", cfg.desired_fps);
+    config_file_emit_single(emitter, "wait-for-vsync", cfg.wait_for_debugger);
```
Comment (1whatleytay) [inline src/emulator/config/src/config.cpp]: Travis is complaining about this now. You hooked up wait-for-vsync to wait-for-debugger. :P

**Comment by 1whatleytay** [comment]
I think everything is good! Some conflicts and then we can merge.

**Comment by pent0** [comment]
Please rebase to master. Sync rendering is removed

File: src/emulator/config/src/config.cpp [hunk 7]
```diff
@@ -370,6 +382,8 @@ ExitCode init_config(Config &cfg, int argc, char **argv, const Root &root_paths)
         LOG_INFO("log-exports: {}", cfg.log_exports);
         LOG_INFO("log-active-shaders: {}", cfg.log_active_shaders);
         LOG_INFO("log-uniforms: {}", cfg.log_uniforms);
+        LOG_INFO("fps-limit-enabled: {}", cfg.fps_limit);
```
Comment (VelocityRa) [inline src/emulator/config/src/config.cpp]:
```suggestion
        LOG_INFO("fps-limit: {}", cfg.fps_limit);
```

File: src/emulator/config/src/config.cpp [hunk 8]
```diff
@@ -370,6 +382,8 @@ ExitCode init_config(Config &cfg, int argc, char **argv, const Root &root_paths)
...
+        LOG_INFO("vsync-enable: {}", cfg.wait_for_vsync);
```
Comment (VelocityRa) [inline src/emulator/config/src/config.cpp]: missing `wait-for-vsync` log

File: src/emulator/main.cpp [hunk 9]
```diff
@@ -154,6 +155,11 @@ int main(int argc, char *argv[]) {
         return RendererInitFailed;
 
     while (handle_events(host)) {
+        std::chrono::steady_clock::time_point frame_start, frame_end;
+
+        if (host.cfg.fps_limit){
```
Comment (VelocityRa) [inline src/emulator/main.cpp]:
```suggestion
        if (host.cfg.fps_limit) {
```

File: src/emulator/main.cpp [hunk 10]
```diff
@@ -154,6 +155,11 @@ int main(int argc, char *argv[]) {
         return RendererInitFailed;
 
     while (handle_events(host)) {
+        std::chrono::steady_clock::time_point frame_start, frame_end;
```
Comment (VelocityRa) [inline src/emulator/main.cpp]: split this to two lines, one for each declaration

File: src/emulator/main.cpp [hunk 11]
```diff
@@ -168,6 +174,18 @@ int main(int argc, char *argv[]) {
         gui::draw_end(host.window.get());
 
         host.display.condvar.notify_all();
+        if (host.cfg.fps_limit) {
+            constexpr double full_second = 1000000.0; //microseconds
```
Comment (VelocityRa) [inline src/emulator/main.cpp]:
```suggestion
            constexpr double full_second = 1000000.0; // microseconds
```

**Comment by 1whatleytay** [comment]
I think this is mergeable as soon as conflicts are fixed :)

**Comment by 1whatleytay** [comment]
Any updates? :D

**Comment by Zangetsu38** [comment]
Outdated

**Comment by Foxlum** [comment]
This looks pretty outdated, and we now have a v-sync option in Vita3k, but no FPS limiter option, but it is very rare you'd want to limit the FPS of a game unless your playing way beyond the 30 or 60 fps cap of most games.
